ls1intum/Artemis
# Update C template and introduce programming language features

### Checklist

### Motivation and Context


The C template was using a quite outdated version of the tester and had some flaws related to potential vulnerabilities.
As we now support a quite large range of programming languages, managing all their features got a bit annoying. This is now simplified in the [ProgrammingLanguageFeatures](https://github.com/ls1intum/Artemis/pull/2150/files#diff-1981327ba5eecfbc95db67961cf2de049944e582c9682aff96d5c9be7cc80d6c).

### Description

I've updated the C template to fix those vulnerabilities and fixed those tests.

Also, the [ProgrammingLanguageFeatures](https://github.com/ls1intum/Artemis/pull/2150/files#diff-1981327ba5eecfbc95db67961cf2de049944e582c9682aff96d5c9be7cc80d6c) now contain the configuration regarding the features of a programming language (e.g. if plagiarism check/sequential test runs/etc. are supported).
The information is present in the server which publishes it to the client using a [ProgrammingLanguageFeatureContributor](https://github.com/ls1intum/Artemis/blob/0cc06eafd85dc46d332d9ee8f53d9205f6801bb5/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingLanguageFeatureContributor.java). The client uses this information to adjust the import/create dialogs (e.g. by only showing a VHDL exercise in Bamboo-setups, as Jenkins does not support VHDL).

### Steps for Testing

1. Create a new C exercises 
2. The solution should build successfully 
3. Play around with the configuration dialog when creating a new programming exercise: The settings should change (e.g. the package name is only present for Java and Kotlin exercises).
4. Verify that all programming exercises are supported in a Bamboo-environment (TS1/TS3) and that a Jenkins-environment (TS2) does only support Java, Python, C and Haskell.

![Peek 2020-10-21 21-55](https://user-images.githubusercontent.com/5084100/96775854-6bfcc480-13e8-11eb-98dd-dc4d6449e432.gif)
